DigitalLearningSolutions.Data/Services/CourseContentService.cs

```diff
@@ -80,18 +80,55 @@ GROUP BY
             ).FirstOrDefault();
         }
 
-        public int GetProgressId(int candidateId, int customisationId)
+        public int? GetOrCreateProgressId(int candidateId, int customisationId, int centreId)
         {
-            // TODO HEEDLS-202: change QueryFirstOrDefault to creating progress record if not found
-            return connection.QueryFirstOrDefault<int>(
-                @"SELECT ProgressId
-                        FROM Progress
-                        WHERE CandidateID = @candidateId
-                          AND CustomisationID = @customisationId",
-                new { candidateId, customisationId }
+            if (DoesProgressExist(candidateId, customisationId))
+            {
+                return GetProgressId(candidateId, customisationId);
+            }
+
+            var errorCode = connection.QueryFirst<int>(
+                @"uspCreateProgressRecord_V3",
+                new
+                {
+                    candidateId,
+                    customisationId,
+                    centreId,
+                    EnrollmentMethodID = 1,
+                    EnrolledByAdminID = 0
+                },
+                commandType: CommandType.StoredProcedure
             );
-        }
 
+            switch (errorCode)
+            {
+                case 0:
+                    return GetProgressId(candidateId, customisationId);
+                case 1:
+                    logger.LogError(
+                        "Not enrolled candidate on course as progress already exists. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;
+                case 100:
+                    logger.LogError(
+                        "Not enrolled candidate on course as customisation id doesn't match centre id. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;
+                case 101:
+                    logger.LogError(
+                        "Not enrolled candidate on course as candidate id doesn't match centre id. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;
+                default:
+                    logger.LogError(
+                        "Not enrolled candidate on course as stored procedure failed. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");

```

Comment:
Probably worth also logging the error code and noting in the message that it's an unknown error code